docs/proposals/emptydir-pv.md

```diff
@@ -0,0 +1,151 @@
+* Every pod referencing this PV must land on the same node (this concept has been referred to as "data gravity").
+* Once this PV is bound to a node, it cannot get unbound till a human intervenes or the node is deleted
+* The PV must require low user interaction. It should be provisionable dynamically.
+* The directory backing the PV must get deleted when the PV is deleted.
+
+## Non goals
+* A local storage abstraction that survives extended node outages
+* Snapshotting or simply copying the local volume to another node
+
+## A new PV storage source
+
+As described above, the storage source for the PV must be a local volume. While we already allow hostPath as a PersistentVolumeSource, it is insufficient because:
+* user needs to specify the exact path, thus requires knowledge of the filesystem
+* hard to dynamically provision a hostpath
+* hostpath volumes are not cleaned up
+
+The `emptydir` abstraction solves these problems:
+* it is allocated in `/var/lib/kubelet/resource/uid/`, a location that is predictable but unknow till `resource` creation time
+* The node takes care of allocation, a dynamic provisioner just needs to create the PV
+* it is cleaned up when the pod dies (or in this case when the PV is deleted)
+
+We can create a new [PersistetVolumeSource](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L240) for EmptyDir. Users will request storage by creating PersistetVolumeClaims. If the PVC isn't bound to a PV, one will be created dynamically. The user can't control which node the storage for this PV comes from, or where in the node's filesystem it is allocated. Users can also hand create PVs bound to specific nodes (TODO: elaborate on QoS and what happens if there's no space on the node).
+
+## Scheduling constraints for data gravity
+
+The scheduler will schedule the pod paying attention to any referenced PVs. If an empty dir PV attached to an existing node is found, that trumps all scheduler predicates. Otherwise the scheduler makes a scheduling decision taking into account pod cpu/memory requests (and in the future, requested storage), attaches it to a node, and all subsequent scheduling decisions for any pod referencing the same PV evaluate to the same node. The scheduler "Attaches" a pv to a node by writing an annotation (TBD what actually writes this annotation, scheduler or kubelet, scheduler will at least bind the pod to the node). All pods referencing the PV that enter the "Running" state on the give node will be able to write to the volume. If the node enters the `NotReady` state, all pods referencing the PV attached to the node will remain in `Pending` till such time when an admin either removes the annotation, or rewrites it to point to another node. The directory allocated to the emptydir volume will be exported through the status field of the PV, so admins can yank the disk from a NotReady node and rsync data to the same directory on another node before rewriting the annotation.
```

Comment:
Here's a straw man, slightly different from yours:

1. Scheduler puts pod on some host
2. Kubelet waits for volume controller to add labels hostname=assignednode because the PV type is "emptyDir" ("attach")
3. Volume controller observes pod scheduled on assignednode, then sets PV label to assignednode
4. Kubelet starts pod (we need to figure out some place to put creation of the pv dir, since attach is used by the volume controller. Lets ignore this for now.)
5. Kubelets only run pods that have PV type emptyDir if the hostname is assigned to the node.
6. Node NotReady, all pods get deleted pods rescheduled on another node
7. Volume controller (or other controller) copies data from dead node onto new node ("detach")
8. Volume controller (or other controller) changes label on PV to other node (second "attach")

Does labels make 1:1 exclusive harder?  We already *must* have node labels for a lot of reasons (so you can assume every node has a unique label) including hard pod affinity, so it seems parsimonious to keep using label based scheduling.